Makefile

```diff
@@ -979,22 +979,26 @@ $(BUILD_DIR)/src/overlays/%_reloc.o: $(BUILD_DIR)/spec
 	$(POSTPROCESS_OBJ) $(@:.o=.s)
 	$(AS) $(ASFLAGS) $(@:.o=.s) -o $@
 
-$(BUILD_DIR)/assets/%.inc.c: assets/%.png
-	$(N64TEXCONV) $(subst .,,$(suffix $*)) "$(findstring u32,$(subst .,,$(suffix $(basename $*))))" $< $@ $(@:.inc.c=.pal.inc.c)
-
-$(BUILD_DIR)/assets/%.inc.c: $(EXTRACTED_DIR)/assets/%.png
-	$(N64TEXCONV) $(subst .,,$(suffix $*)) "$(findstring u32,$(subst .,,$(suffix $(basename $*))))" $< $@ $(@:.inc.c=.pal.inc.c)
+# Assets from assets/
 
-$(BUILD_DIR)/assets/%.bin.inc.c: assets/%.bin
-	$(BIN2C) -t 1 $< $@
+$(BUILD_DIR)/assets/%.inc.c: assets/%.png
+	tools/assets/build_from_png/build_from_png $< $(dir $@) assets/$(dir $*) $(wildcard $(EXTRACTED_DIR)/assets/$(dir $*))
 
-$(BUILD_DIR)/assets/%.bin.inc.c: $(EXTRACTED_DIR)/assets/%.bin
+$(BUILD_DIR)/assets/%.u8.inc.c: assets/%.u8.bin
 	$(BIN2C) -t 1 $< $@
 
-$(BUILD_DIR)/assets/%.jpg.inc.c: assets/%.jpg
+$(BUILD_DIR)/assets/%.u64.jpg.inc.c: assets/%.u64.jpg
 	$(N64TEXCONV) JFIF "" $< $@
 
-$(BUILD_DIR)/assets/%.jpg.inc.c: $(EXTRACTED_DIR)/assets/%.jpg
+# Assets from extracted/
+
+$(BUILD_DIR)/assets/%.inc.c: $(EXTRACTED_DIR)/assets/%.png
+	tools/assets/build_from_png/build_from_png $< $(dir $@) $(wildcard assets/$(dir $*)) $(EXTRACTED_DIR)/assets/$(dir $*)
```

Comment:
imo it's cleaner to let the makefile do the relevant string manipulations on the filename rather than do it in C in the tool, but I haven't prototyped what this would look like so maybe there's something I'm missing

Reply:
After looking into this a bit, I think it's ok to leave it as-is, a makefile-centric approach would probably involve a bunch of bash scripts or `$(shell ...)` calls anyway